### PR TITLE
Also match friendly sessions based on the buffer's ns form

### PR DESCRIPTION
**Pull request by vemv**
> Fixes https://github.com/clojure-emacs/cider/issues/3419

Sample use case:

* Start a REPL on the Haystack project
  * REPL / interactions work, as usual
* Switch to the Orchard project, visit a ns file
  * REPL / interactions work, because the ns form matches with Haystack's ns list
  * (Orchard is a direct dependency of Haystack)
* Switch to an unrelated project
  * REPL / interactions rightfully won't work (`No linked CIDER sessions`), because the ns form won't match with Haystack's ns list

The "ns list" is basically:

* `(all-ns)`, computed once
* `cider-repl-ns-cache`, i.e. the changed namespaces as conveyed by `track-state`.

Cheers - V

### CODE REVIEW

File: cider-connection.el [hunk 1]
```diff
@@ -615,64 +615,6 @@ REPL defaults to the current REPL."
 
 (declare-function cider-classpath-entries "cider-client")
 
-(defun cider--sesman-friendly-session-p (session &optional debug)
```
Comment (vemv) [inline cider-connection.el:618]: Because the new impl uses misc variables/defuns like cider-repl-ns-cache, cider-sync-request:ns-list, I had to move it, keeping the linter happy.

tbh it makes sense to me that if this code now uses more repl-based capabilities, it's defined in cider-repl.el.

An alternative possibly being `declare-function`?

File: cider-repl.el [hunk 2]
```diff
@@ -1746,6 +1746,79 @@ constructs."
...
+      (let* ((classpath (or (process-get proc :cached-classpath)
+                            (let ((cp (with-current-buffer repl
+                                        (cider-classpath-entries))))
+                              (process-put proc :cached-classpath cp)
+                              cp)))
+             (ns-list (or (process-get proc :all-namespaces)
+                          (let ((ns-list (with-current-buffer repl
+                                           (cider-sync-request:ns-list))))
+                            (process-put proc :all-namespaces ns-list)
+                            ns-list)))
```
Comment (vemv) [inline cider-repl.el:1772]: This section is new.

File: cider-repl.el [hunk 3]
```diff
@@ -1746,6 +1746,79 @@ constructs."
...
+                        translated))
+            (when-let ((ns (condition-case nil
+                               (substring-no-properties (cider-current-ns :no-default))
+                             (error nil))))
+              ;; if the ns form matches with a ns of all runtime namespaces, we can consider the buffer to match
+              ;; (this is a bit lax, but also quite useful)
+              (with-current-buffer repl
+                (or (when cider-repl-ns-cache ;; may be nil on repl startup
+                      (member ns (nrepl-dict-keys cider-repl-ns-cache)))
+                    (member ns ns-list))))
```
Comment (vemv) [inline cider-repl.el:1804]: This section is new.

**Review by vemv — COMMENTED** [review]
The diff might be a little confusing, but luckily, it boils to two additions that I've marked as multi-line comments.

File: cider-repl.el (hunk 3 above)
Comment (bbatsov) [reply cider-repl.el:1804]: Looks good to me. At some point it'd be nice to comment all the cases for a friendly session as the function is now huge and it's kind of hard to follow.

File: cider-connection.el (hunk 1 above)
Comment (bbatsov) [reply cider-connection.el:618]: I'm not a big fan of `declare-function`, so let's keep it moved.

**Comment by bbatsov** [comment]
Looks good to me now.
